proposals/text/5629-file-similarity-retriever.md

```diff
@@ -0,0 +1,186 @@
+- Title: Adding FileSimilarityRetriever to haystack
+- Decision driver: @elundaeva
+- Start Date: 2023-08-28
+- Proposal PR: https://github.com/deepset-ai/haystack/pull/5629
+- Github Issue or Discussion: some links available in the "Detailed design" section below
+
+# Summary
+
+The retriever takes a metadata aggregation key (e.g. "file_id") as query, searches for all documents from that file in the doc store and then performs a query for each document to find similar documents for each. Then these search results for each document are aggregated to produce a list of similar files.
+
+# Basic example
+
+The FileSimilarityRetriever would be instantiated as follows:
+
+``` python
+
+	  retriever = FileSimilarityRetriever(
+	      document_store = ElasticSearchDocumentStore,
+	      primary_retriever = EmbeddingRetriever, # defined separately, see full pipeline example below
+        top_k=30,
+	      file_aggregation_key = "file_id",
+        max_num_queries = 50
+	  )
+```
+
+And here is an example of how the node would work in the context of a full pipeline:
+
+``` yaml
+
+version: '1.19.0'
+name: 'FileSim'
+
+components:
+  - name: DocumentStore
+    type: DeepsetCloudDocumentStore
+    params:
+      similarity: cosine
+      embedding_dim: 768
+      return_embedding: false
+
+  - name: EmbeddingRetriever
+    type: EmbeddingRetriever
+    params:
+      document_store: DocumentStore
+      embedding_model: sentence-transformers/all-mpnet-base-v2
+      max_seq_len: 400
+      batch_size: 32
+      model_format: sentence_transformers
+      top_k: 50
+  - name: TextConverter
+    type: TextConverter
+  - name: FileSimilarityRetriever
+    type: FileSimilarityRetriever
+    params:
+      document_store: DocumentStore
+      primary_retriever: EmbeddingRetriever
+      top_k: 30
+      file_aggregation_key: file_id
+      max_num_queries: 50
+
+pipelines:
+  - name: query
+    nodes:
+      - name: FileSimilarityRetriever
+        inputs: [Query]
+
+  - name: indexing
+    nodes:
+      - name: TextConverter
+        inputs: [File]
+
+      - name: EmbeddingRetriever
+        inputs: [TextConverter]
+
+      - name: DocumentStore
+        inputs: [EmbeddingRetriever]
+
+```
+
+# Basic example - updated
+
+When implemented as two sub-components (see the section **Detailed design - updated** below), this retriever would be instantiated as follows:
+
+``` python
+
+	  DocumentFetcher = DocumentFetcher(
+	      document_store = ElasticSearchDocumentStore,
+	      file_aggregation_key = "file_id",
+        max_num_docs = 50
+	  )
+
+    >>> then one or two retrievers here >>>
+
+    DocumentAggregator = DocumentAggregator(
+        top_k = 5,
+        file_aggregation_key = "file_id",
+        output = "top_document", # This new param is explained in Detailed design section below
+        keep_original_score = True
+    )
+
```

Comment:
I'm a little confused by the full pipeline flow here. Could you provide code examples for the Retriever(s) and how they interact with the `DocumentAggregator`? Will these components be passed to a wrapper node like `FileSimilarityRetriever`?

Reply:
I think how we package it is still an open question (related to [this comment](https://github.com/deepset-ai/haystack/pull/5629#issuecomment-1705258037)) - whether it makes sense to make a wrapper node already now or if things change drastically with Haystack 2.0 it could be done later to avoid double work 🤔